<div>
<img src="https://www.ul.ie/themes/custom/ul/logo.jpg" />
</div>

#**MSc in Artificial Intelligence and Machine Learning**
##CS6271 - Evolutionary Algorithms and Humanoid Robotics 2024
### Kaggle Competition


Module Leader: Conor Ryan

Developer: Allan De Lima

Link to access the competition: https://www.kaggle.com/competitions/cs-6271-2024-5-final-project

Link to join the competition: https://www.kaggle.com/t/017d7291cce043fcb79fb087b6a78437



## Introduction

Predict whether a pacient has diabetes. This is a shorter version of the also known as "Diabetes" dataset.

In [1]:
# Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Dataset

**Class:**

output: 1, 0.


**Listing of features:**

HighBP

HighChol

CholCheck

BMI

Smoker

Stroke

HeartDiseaseorAttack

PhysActivity

Fruits

Veggies

HvyAlcoholConsump

AnyHealthcare

NoDocbcCost

GenHlth

MentHlth

PhysHlth

DiffWalk

Sex

Age

Education

Income

### Load the dataset

In [2]:
# Suppressing Warnings:
import warnings
warnings.filterwarnings("ignore")

In [3]:
## mount your Google drive
# 1) run this cell
# 2) sign in
# 3) verify your drive is mounted

from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


Clone the GRAPE repository at first because the dataset to be used is already there.

In [4]:
import os
# Get the library from our BDS research Group
# copy the path from your drive
PATH = '/content/drive/MyDrive/grape/'

# check if 'grape' already exists
if os.path.exists(PATH):
    print('grape directory already exists')
else:
    %cd /content/drive/MyDrive/
    !git clone https://github.com/bdsul/grape.git
    print('Cloning grape in your Drive')

# change directory to 'grape'
%cd /content/drive/MyDrive/grape/

grape directory already exists
/content/drive/MyDrive/grape


Now you have a grape folder in your Drive account.

Upload the files training.csv and test.csv from the competition to the folder grape/datasets in your Drive before running the next cells.

### Train set

In [120]:
train_file = 'datasets/training.csv'

In [121]:
# load train set
df_train = pd.read_csv(PATH+train_file)
df_train.head()

HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  HeartDiseaseorAttack  \
0     0.0       0.0        1.0  27.0     0.0     0.0                   0.0   
1     1.0       1.0        1.0  35.0     0.0     0.0                   1.0   
2     1.0       1.0        1.0  24.0     0.0     0.0                   0.0   
3     0.0       0.0        1.0  24.0     1.0     0.0                   0.0   
4     1.0       0.0        1.0  41.0     0.0     0.0                   0.0   

   PhysActivity  Fruits  Veggies  ...  NoDocbcCost  GenHlth  MentHlth  \
0           1.0     1.0      1.0  ...          0.0      1.0       1.0   
1           1.0     1.0      1.0  ...          0.0      2.0       0.0   
2           1.0     1.0      1.0  ...          0.0      2.0       0.0   
3           1.0     0.0      1.0  ...          0.0      2.0       0.0   
4           0.0     0.0      0.0  ...          0.0      3.0       3.0   

   PhysHlth  DiffWalk  Sex   Age  Education  Income  output  
0       0.0       0.0  0.0   7.0        6.0     8.0     0.0  
1       0.0       0.0  1.0   9.0        6.0     8.0     0.0  
2       0.0       1.0  0.0  13.0        5.0     4.0     0.0  
3       0.0       0.0  1.0   7.0        4.0     6.0     0.0  
4       0.0       0.0  0.0   8.0        5.0     7.0     0.0  

[5 rows x 22 columns]

In [122]:
df_train.describe()

HighBP     HighChol    CholCheck          BMI       Smoker  \
count  5042.000000  5042.000000  5042.000000  5042.000000  5042.000000   
mean      0.513883     0.493653     0.972035    29.447441     0.473225   
std       0.499857     0.500009     0.164889     7.080019     0.499332   
min       0.000000     0.000000     0.000000    14.000000     0.000000   
25%       0.000000     0.000000     1.000000    25.000000     0.000000   
50%       1.000000     0.000000     1.000000    28.000000     0.000000   
75%       1.000000     1.000000     1.000000    33.000000     1.000000   
max       1.000000     1.000000     1.000000    98.000000     1.000000   

            Stroke  HeartDiseaseorAttack  PhysActivity       Fruits  \
count  5042.000000           5042.000000   5042.000000  5042.000000   
mean      0.050377              0.133082      0.728877     0.616025   
std       0.218743              0.339697      0.444583     0.486400   
min       0.000000              0.000000      0.000000     0.000000   
25%       0.000000              0.000000      0.000000     0.000000   
50%       0.000000              0.000000      1.000000     1.000000   
75%       0.000000              0.000000      1.000000     1.000000   
max       1.000000              1.000000      1.000000     1.000000   

           Veggies  ...  NoDocbcCost      GenHlth     MentHlth     PhysHlth  \
count  5042.000000  ...  5042.000000  5042.000000  5042.000000  5042.000000   
mean      0.801666  ...     0.085879     2.702102     3.617017     5.195954   
std       0.398785  ...     0.280213     1.106379     7.997166     9.538762   
min       0.000000  ...     0.000000     1.000000     0.000000     0.000000   
25%       1.000000  ...     0.000000     2.000000     0.000000     0.000000   
50%       1.000000  ...     0.000000     3.000000     0.000000     0.000000   
75%       1.000000  ...     0.000000     3.000000     2.000000     5.000000   
max       1.000000  ...     1.000000     5.000000    30.000000    30.000000   

          DiffWalk          Sex          Age    Education       Income  \
count  5042.000000  5042.000000  5042.000000  5042.000000  5042.000000   
mean      0.207061     0.447441     8.376041     4.958548     5.852836   
std       0.405240     0.497279     2.920253     1.018217     2.131317   
min       0.000000     0.000000     1.000000     1.000000     1.000000   
25%       0.000000     0.000000     7.000000     4.000000     4.000000   
50%       0.000000     0.000000     9.000000     5.000000     6.000000   
75%       0.000000     1.000000    10.000000     6.000000     8.000000   
max       1.000000     1.000000    13.000000     6.000000     8.000000   

            output  
count  5042.000000  
mean      0.356010  
std       0.478866  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       1.000000  
max       1.000000  

[8 rows x 22 columns]

In [123]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
def scal(data,val):

  X = data.copy()
  # Columns for each scaling method
  standard_cols = ['BMI', 'Age', 'MentHlth', 'PhysHlth','GenHlth', 'Education', 'Income']
  if val == "train":
    scaler.fit(X[standard_cols])
    X[standard_cols] = scaler.transform(X[standard_cols])
  else:
    X[standard_cols] = scaler.transform(X[standard_cols])
  return X




In [124]:
X_train = df_train.copy()
X_train.drop(['output'], axis=1, inplace=True)
y_train = df_train['output'].to_numpy()
X_train = scal(X_train,"train")
X_train = X_train[['GenHlth','HighBP' , 'HighChol'  ,'BMI' , 'DiffWalk' , 'Age' , 'HeartDiseaseorAttack'  , 'PhysHlth',"Stroke","CholCheck","Smoker","MentHlth","Veggies","Income"]]
print(X_train)

       GenHlth  HighBP  HighChol       BMI  DiffWalk       Age  \
0    -1.538597     0.0       0.0 -0.345717       0.0 -0.471253   
1    -0.634658     1.0       1.0  0.784335       0.0  0.213687   
2    -0.634658     1.0       1.0 -0.769487       1.0  1.583567   
3    -0.634658     0.0       0.0 -0.769487       0.0 -0.471253   
4     0.269281     1.0       0.0  1.631875       0.0 -0.128783   
...        ...     ...       ...       ...       ...       ...   
5037  0.269281     1.0       1.0  0.643079       1.0  1.241097   
5038 -0.634658     1.0       1.0 -0.345717       0.0 -0.128783   
5039  0.269281     0.0       0.0 -1.052000       0.0  0.898627   
5040  1.173221     1.0       1.0  0.360566       1.0  1.583567   
5041  0.269281     1.0       0.0 -0.769487       0.0  1.583567   

      HeartDiseaseorAttack  PhysHlth  Stroke  CholCheck  Smoker  MentHlth  \
0                      0.0 -0.544774     0.0        1.0     0.0 -0.327276   
1                      1.0 -0.544774     0.0        1

In [125]:
print(y_train[0:5]) #print head

[0. 0. 0. 0. 0.]


### Test set

In [126]:
test_file = 'datasets/test.csv'

In [127]:
# load test set
df_test = pd.read_csv(PATH+test_file)
df_test.head()

HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  HeartDiseaseorAttack  \
0     1.0       1.0        1.0  28.0     1.0     0.0                   1.0   
1     1.0       0.0        1.0  27.0     1.0     0.0                   1.0   
2     1.0       1.0        1.0  27.0     0.0     0.0                   0.0   
3     0.0       0.0        1.0  27.0     0.0     0.0                   0.0   
4     1.0       1.0        1.0  32.0     1.0     0.0                   1.0   

   PhysActivity  Fruits  Veggies  ...  AnyHealthcare  NoDocbcCost  GenHlth  \
0           1.0     0.0      0.0  ...            0.0          0.0      4.0   
1           0.0     1.0      1.0  ...            1.0          0.0      3.0   
2           1.0     1.0      1.0  ...            1.0          0.0      1.0   
3           1.0     0.0      1.0  ...            1.0          0.0      3.0   
4           0.0     0.0      0.0  ...            1.0          0.0      5.0   

   MentHlth  PhysHlth  DiffWalk  Sex   Age  Education  Income  
0       2.0       2.0       1.0  1.0   8.0        4.0     1.0  
1       0.0      14.0       0.0  0.0  11.0        6.0     6.0  
2       3.0       0.0       0.0  1.0  12.0        6.0     6.0  
3       0.0       0.0       0.0  0.0   9.0        6.0     3.0  
4       3.0       6.0       1.0  0.0  10.0        5.0     6.0  

[5 rows x 21 columns]

In [128]:
df_test.describe()

HighBP      HighChol     CholCheck           BMI        Smoker  \
count  95804.000000  95804.000000  95804.000000  95804.000000  95804.000000   
mean       0.508330      0.484239      0.970628     29.262797      0.461442   
std        0.499933      0.499754      0.168849      6.972092      0.498514   
min        0.000000      0.000000      0.000000     12.000000      0.000000   
25%        0.000000      0.000000      1.000000     25.000000      0.000000   
50%        1.000000      0.000000      1.000000     28.000000      0.000000   
75%        1.000000      1.000000      1.000000     32.000000      1.000000   
max        1.000000      1.000000      1.000000     98.000000      1.000000   

             Stroke  HeartDiseaseorAttack  PhysActivity        Fruits  \
count  95804.000000          95804.000000  95804.000000  95804.000000   
mean       0.053839              0.125329      0.724427      0.621352   
std        0.225701              0.331093      0.446805      0.485053   
min        0.000000              0.000000      0.000000      0.000000   
25%        0.000000              0.000000      0.000000      0.000000   
50%        0.000000              0.000000      1.000000      1.000000   
75%        0.000000              0.000000      1.000000      1.000000   
max        1.000000              1.000000      1.000000      1.000000   

            Veggies  ...  AnyHealthcare   NoDocbcCost       GenHlth  \
count  95804.000000  ...   95804.000000  95804.000000  95804.000000   
mean       0.798161  ...       0.953353      0.088973      2.701891   
std        0.401375  ...       0.210883      0.284707      1.106275   
min        0.000000  ...       0.000000      0.000000      1.000000   
25%        1.000000  ...       1.000000      0.000000      2.000000   
50%        1.000000  ...       1.000000      0.000000      3.000000   
75%        1.000000  ...       1.000000      0.000000      3.000000   
max        1.000000  ...       1.000000      1.000000      5.000000   

           MentHlth      PhysHlth      DiffWalk           Sex           Age  \
count  95804.000000  95804.000000  95804.000000  95804.000000  95804.000000   
mean       3.488654      5.155662      0.219573      0.447841      8.362761   
std        7.819002      9.550624      0.413960      0.497275      2.959191   
min        0.000000      0.000000      0.000000      0.000000      1.000000   
25%        0.000000      0.000000      0.000000      0.000000      7.000000   
50%        0.000000      0.000000      0.000000      0.000000      9.000000   
75%        2.000000      5.000000      0.000000      1.000000     11.000000   
max       30.000000     30.000000      1.000000      1.000000     13.000000   

          Education        Income  
count  95804.000000  95804.000000  
mean       4.977642      5.846739  
std        1.012794      2.136366  
min        1.000000      1.000000  
25%        4.000000      4.000000  
50%        5.000000      6.000000  
75%        6.000000      8.000000  
max        6.000000      8.000000  

[8 rows x 21 columns]

In [129]:
X1_test = df_test.copy()
X1_test = scal(X1_test , "test")
X1_test = X1_test[['GenHlth','HighBP' , 'HighChol'  ,'BMI' , 'DiffWalk' , 'Age' , 'HeartDiseaseorAttack'  , 'PhysHlth',"Stroke","CholCheck","Smoker","MentHlth","Veggies","Fruits"]]
print(X1_test)

        GenHlth  HighBP  HighChol       BMI  DiffWalk       Age  \
0      1.173221     1.0       1.0 -0.204461       1.0 -0.128783   
1      0.269281     1.0       0.0 -0.345717       0.0  0.898627   
2     -1.538597     1.0       1.0 -0.345717       0.0  1.241097   
3      0.269281     0.0       0.0 -0.345717       0.0  0.213687   
4      2.077160     1.0       1.0  0.360566       1.0  0.556157   
...         ...     ...       ...       ...       ...       ...   
95799  0.269281     1.0       1.0 -0.628230       1.0  1.583567   
95800  1.173221     1.0       0.0 -1.193257       0.0 -2.526073   
95801 -0.634658     0.0       0.0  1.066848       0.0 -1.841133   
95802  0.269281     1.0       1.0  0.219309       0.0  0.898627   
95803 -0.634658     1.0       1.0 -0.486974       0.0  1.241097   

       HeartDiseaseorAttack  PhysHlth  Stroke  CholCheck  Smoker  MentHlth  \
0                       1.0 -0.335082     0.0        1.0     1.0 -0.202219   
1                       1.0  0.923067  

You will need to prepare both training and test datasets before working with a Machine Learning method.

Consider you need to use some encoding method with categorical data.

You are free to use any other pre-processing ideas.

In [ ]:
#Include your code here

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_train, y_train, test_size=0.3, random_state=42)

Convert the datasets to NumPy to easily use them.

In [130]:
# data features
X_train = X_train.to_numpy()
#X_test = X_test.to_numpy()


In [131]:
f_test = X1_test

## GRAPE

<div>
<img src="https://drive.google.com/uc?export=view&id=1hw43Oi3lGTCkspQ0ged2bZB8q2EpcPhz" width="150"/>
</div>

GRammatical Algorithms in Python for Evolution (GRAPE)


In [132]:
!pip install deap

import grape
import algorithms

from os import path
from deap import creator, base, tools
import random
import csv

You can import functions to be used with your grammar from [functions.py](https://github.com/UL-BDS/grape/blob/main/functions.py) on GRAPE repository and / or you can define your own functions.

In [133]:
from functions import add, sub, mul, pdiv, psqrt, plog, and_, or_, not_, less_than_or_equal, greater_than_or_equal,if_

'heartDisease.bnf' is a grammar used for another problem just to check if everything is working well.

Write your own grammar in a text file and save it in your Drive account.

Put the whole address on GRAMMAR_FILE and print to check it.

In [134]:
#GRAMMAR_FILE = '/content/drive/MyDrive/data/example.bnf' #put the whole address of your own grammar and remove the # in the beginning of this line
GRAMMAR_FILE = 'heartDisease.bnf' #remove this line when you are using your own grammar

#f = open(GRAMMAR_FILE, "r") #remove the # in the beginning of this line when you are using your own grammar
f = open("grammars/" + GRAMMAR_FILE, "r") #remove this line when you are using your own grammar
print(f.read())
f.close()


<log_op> ::= <conditional_branches> | and_(<log_op>,<log_op>) | or_(<log_op>,<log_op>) | not_(<log_op>) | <boolean_feature>
<conditional_branches> ::= less_than_or_equal(<num_op>,<num_op>) | greater_than_or_equal(<num_op>, <num_op>) | if_(<log_op>,<conditional_branches>,<conditional_branches>)
<num_op>   ::= add(<num_op>,<num_op>) | sub(<num_op>,<num_op>) | mul(<num_op>,<num_op>) | pdiv(<num_op>,<num_op>) | <nonboolean_feature>
<boolean_feature> ::= x[1]|x[2]|x[4]|x[6]|x[8]|x[9]|x[10]|x[12]
<nonboolean_feature> ::= x[0]|x[3]|x[5]|x[7]|x[11]|x[13]|<constant>
<constant> ::= -2 | -1.5 | -1 | -0.5 | 0 | 0.5 | 1 | 1.5 | 2 |<precise_constant>
<precise_constant> ::= <c><c>.<c><c>
<c> ::= 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9


Run the following cell to put your grammar on the class Grammar.

In [135]:
#BNF_GRAMMAR = grape.Grammar(GRAMMAR_FILE) #remove the # in the beginning of this line when you are using your own grammar
BNF_GRAMMAR = grape.Grammar(path.join("grammars", GRAMMAR_FILE)) #remove this line when you are using your own grammar

In [136]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


The fitness function here is the percentage of outputs wrongly predicted.

You can write your own fitness function if you prefer.

In [137]:
import numpy as np

def fitness_eval(individual, points):
    max_repetitions = 3             # Maximum allowed repetitions for a feature
    length_penalty_factor = 0.0005   # Length penalty factor (smaller for less impact)
    repetitive_penalty_factor = 0.5  # Lower repetitive penalty for key features
    accuracy_weight = 1.5            # Emphasize accuracy in fitness
    x = points[0]
    y = points[1]

    if individual.invalid:
        return 1000,

    try:
        # Evaluate the individual's phenotype to generate predictions
        pred = eval(individual.phenotype)
    except (FloatingPointError, ZeroDivisionError, OverflowError, MemoryError):
        return 1000,

    assert np.isrealobj(pred)

    try:
        # Binary classification with an adjusted threshold
        Y_class = [1 if pred[i] > 0.5 else 0 for i in range(len(y))]
    except (IndexError, TypeError):
        return 1000,

    # Accuracy component: Encourage correct classifications
    accuracy = np.mean(np.equal(y, Y_class))
    accuracy_penalty = 1 - accuracy  # Higher accuracy lowers this component

    # Count feature occurrences in phenotype for penalties
    phenotype_str = str(individual.phenotype)
    feature_counts = {f"x[{i}]": phenotype_str.count(f"x[{i}]") for i in range(len(x[0]))}

    # Penalty for overly repetitive features
    repetitive_penalty = sum(
        (count - max_repetitions) * repetitive_penalty_factor
        for count in feature_counts.values() if count > max_repetitions
    )

    # Length penalty based on the expression size


    # Minimized fitness calculation: Accuracy weighted with penalties
    fitness = (accuracy_penalty ) + repetitive_penalty

    return (fitness,)


To use properly the fitness function above with GRAPE, the features must be in the lines, and the samples must be in the columns, so if your data is not like that, you need to transpose the matrix.

Take a look at the print. If you run this cell two times, the matrix will be transposed again and will not work properly.

In [138]:
X_train = np.transpose(X_train)
#X_test = np.transpose(X_test)

print('Training (X,Y):\t', X_train.shape, y_train.shape)
#print('Test (X):\t', X_test.shape)

Training (X,Y):	 (14, 5042) (5042,)


Set the Grammatical Evolution parameters.

Make sure you set a random seed just in case we need to re-run your experiments.

In [142]:
POPULATION_SIZE = 1000
MAX_GENERATIONS = 250
P_CROSSOVER = 0.8
P_MUTATION = 0.03
ELITE_SIZE = 4
HALL_OF_FAME_SIZE = 5

TOURNAMENT_SIZE = 6
RANDOM_SEED = 42
random.seed(RANDOM_SEED)

CODON_CONSUMPTION = 'lazy'
GENOME_REPRESENTATION = 'list'
MAX_GENOME_LENGTH = None

MAX_INIT_TREE_DEPTH = 13
MIN_INIT_TREE_DEPTH = 3
MAX_TREE_DEPTH = 60
MAX_WRAPS = 0
CODON_SIZE = 255

REPORT_ITEMS = ['gen', 'invalid', 'avg', 'std', 'min', 'max',
                'best_ind_length', 'avg_length',
                'best_ind_nodes', 'avg_nodes',
                'best_ind_depth', 'avg_depth',
                'avg_used_codons', 'best_ind_used_codons',
                'structural_diversity', 'fitness_diversity',
                'selection_time', 'generation_time']

Create a toolbox.

In [143]:
toolbox = base.Toolbox()

# define a single objective, minimising fitness strategy:
creator.create("FitnessMin", base.Fitness, weights=(-1.0,))

creator.create('Individual', grape.Individual, fitness=creator.FitnessMin)


toolbox.register("populationCreator", grape.sensible_initialisation, creator.Individual)

toolbox.register("evaluate", fitness_eval)

# Tournament selection:
toolbox.register("select", tools.selTournament, tournsize=TOURNAMENT_SIZE)

# Single-point crossover:
toolbox.register("mate", grape.crossover_onepoint)

# Flip-int mutation:
toolbox.register("mutate", grape.mutation_int_flip_per_codon)


In [144]:
# create initial population (generation 0):
population = toolbox.populationCreator(pop_size=POPULATION_SIZE,
                                           bnf_grammar=BNF_GRAMMAR,
                                           min_init_depth=MIN_INIT_TREE_DEPTH,
                                           max_init_depth=MAX_INIT_TREE_DEPTH,
                                           codon_size=CODON_SIZE,
                                           codon_consumption=CODON_CONSUMPTION,
                                           genome_representation=GENOME_REPRESENTATION
                                            )

# define the hall-of-fame object:
hof = tools.HallOfFame(HALL_OF_FAME_SIZE)

# prepare the statistics object:
stats = tools.Statistics(key=lambda ind: ind.fitness.values)
stats.register("avg", np.nanmean)
stats.register("std", np.nanstd)
stats.register("min", np.nanmin)
stats.register("max", np.nanmax)

In [ ]:
for ind in population:
  ind.fitness.values = toolbox.evaluate(ind, points=[X_train, y_train])
  print(ind.fitness.values)

(0.6358563288630669,)
(0.5926525643147265,)
(0.5926525643147265,)
(0.6358563288630669,)
(0.6405515323064093,)
(0.6405515323064093,)
(0.5926525643147265,)
(0.5926525643147265,)
(0.5926525643147265,)
(0.6405515323064093,)
(0.6358563288630669,)
(0.571742897850169,)
(0.6405515323064093,)
(0.5926525643147265,)
(0.6405515323064093,)
(0.571742897850169,)
(0.6358563288630669,)
(0.6358563288630669,)
(0.5926525643147265,)
(0.6405515323064093,)
(0.5926525643147265,)
(0.6405515323064093,)
(0.5926525643147265,)
(0.5926525643147265,)
(0.5926525643147265,)
(0.5926525643147265,)
(0.6358563288630669,)
(0.6405515323064093,)
(0.571742897850169,)
(0.6405515323064093,)
(0.6358563288630669,)
(0.571742897850169,)
(0.571742897850169,)
(0.6358563288630669,)
(0.6358563288630669,)
(0.5926525643147265,)
(0.6405515323064093,)
(0.5926525643147265,)
(0.6530686921795652,)
(0.571742897850169,)
(0.634408407791823,)
(0.26818528862980684,)
(0.5926525643147265,)
(0.24730611572187278,)
(0.648816837450002,)
(0.6530686921795

KeyboardInterrupt: 

Run Grammatical Evolution.

In [145]:
population, logbook = algorithms.ge_eaSimpleWithElitism(population, toolbox, cxpb=P_CROSSOVER, mutpb=P_MUTATION,
                                              ngen=MAX_GENERATIONS, elite_size=ELITE_SIZE,
                                              bnf_grammar=BNF_GRAMMAR,
                                              codon_size=CODON_SIZE,
                                              max_tree_depth=MAX_TREE_DEPTH,
                                              max_genome_length=MAX_GENOME_LENGTH,
                                              points_train=[X_train, y_train],
                                              codon_consumption=CODON_CONSUMPTION,
                                              report_items=REPORT_ITEMS,
                                              genome_representation=GENOME_REPRESENTATION,
                                              stats=stats, halloffame=hof, verbose=False)

gen = 0 , Best fitness = (0.2943276477588259,)
gen = 1 , Best fitness = (0.2943276477588259,) , Number of invalids = 238
gen = 2 , Best fitness = (0.2943276477588259,) , Number of invalids = 118
gen = 3 , Best fitness = (0.29373264577548597,) , Number of invalids = 204
gen = 4 , Best fitness = (0.2935343117810393,) , Number of invalids = 246
gen = 5 , Best fitness = (0.2907576358587862,) , Number of invalids = 202
gen = 6 , Best fitness = (0.2907576358587862,) , Number of invalids = 157
gen = 7 , Best fitness = (0.2907576358587862,) , Number of invalids = 144
gen = 8 , Best fitness = (0.2907576358587862,) , Number of invalids = 127
gen = 9 , Best fitness = (0.2828242760809203,) , Number of invalids = 153
gen = 10 , Best fitness = (0.2816342721142404,) , Number of invalids = 154
gen = 11 , Best fitness = (0.2816342721142404,) , Number of invalids = 181
gen = 12 , Best fitness = (0.2816342721142404,) , Number of invalids = 168
gen = 13 , Best fitness = (0.2816342721142404,) , Number of i

Show the best individual as an expression.

In [146]:
# Best individual
import textwrap
best = hof.items[0].phenotype
print("Best individual: \n","\n".join(textwrap.wrap(best,80)))
print("\nTraining Fitness: ", hof.items[0].fitness.values[0])

Best individual: 
 and_(less_than_or_equal(mul(x[3],-
1),x[0]),or_(x[6],or_(x[8],or_(x[8],and_(x[1],or_(or_(x[6],and_(x[2],or_(x[6],an
d_(x[9],x[1])))),and_(less_than_or_equal(x[5],sub(x[5],sub(x[0],x[3]))),and_(x[1
],less_than_or_equal(mul(x[5],-1.5),x[0])))))))))

Training Fitness:  0.25366917889726304


Define a function to predict values, without comparing to expected outputs.

In [147]:
def predict_train(individual, X , y_test  ):
    x = X

    if individual.invalid == True:
        return np.NaN, # Return NaN if the individual is invalid

    # Evaluate the expression
    try:
      def evaluate_expression(x):
          # Define local variables for the expression context
          X = x
          # Evaluate the expression and handle potential errors
          try:
              pred = eval(individual.phenotype)
          except (TypeError, ValueError, NameError):  # Handle potential errors in expression
              pred = np.nan  # Return NaN for invalid expressions
          return pred  # Explicitly return the prediction

      pred = np.apply_along_axis(evaluate_expression,0,x)
    except (FloatingPointError, ZeroDivisionError, OverflowError,
            MemoryError):
        return np.NaN,

    # Replace NaN values with a suitable default (e.g., -1)
    pred = np.nan_to_num(pred, nan=-1).astype(int)

    y_test = y_test.astype(int)
    pred = pred.astype(int)  # This line is now handled above
    acc = accuracy_score(y_test, pred)


    return pred , acc

In [148]:
def predict_test(individual, X  ):
    x = X

    if individual.invalid == True:
        return np.NaN, # Return NaN if the individual is invalid

    # Evaluate the expression
    try:
      def evaluate_expression(x):
          # Define local variables for the expression context
          X = x
          # Evaluate the expression and handle potential errors
          try:
              pred = eval(individual.phenotype)
          except (TypeError, ValueError, NameError):  # Handle potential errors in expression
              pred = np.nan  # Return NaN for invalid expressions
          return pred  # Explicitly return the prediction

      pred = np.apply_along_axis(evaluate_expression,0,x)
    except (FloatingPointError, ZeroDivisionError, OverflowError,
            MemoryError):
        return np.NaN,

    # Replace NaN values with a suitable default (e.g., -1)
    pred = np.nan_to_num(pred, nan=-1).astype(int)

    #y_test = y_test.astype(int)
    pred = pred.astype(int)  # This line is now handled above
    #acc = accuracy_score(y_test, pred)


    return pred

In [149]:



f_test = np.transpose(f_test)
print(f_test.shape)


(14, 95804)


Predict the classes of the test set.

Make sure you print here in the notebook you will submit to Brightspace the same predictions you used in your best submission to the Kaggle competition.

In [150]:
from sklearn.metrics import accuracy_score
print(hof.items[0].phenotype)
print("Training Fitness: ", hof.items[0].fitness.values[0])
#f_test = f_test.to_numpy()
#f_test = np.transpose(f_test)

print(f_test.shape)
#print(y_test.shape)
y_pred_train , acc = predict_train(hof.items[0], X_train, y_train)
print("Predicted classes of the training set: ", y_pred_train)
print("Accuracy of the training set: ", acc)

y_pred  = predict_test(hof.items[0], f_test)
print("Predicted classes of the test set: ", y_pred  )
data = {"output" : y_pred}
df1 = pd.DataFrame(data)
print(df1.head(100))
df1.to_csv("/content/sample_data/result1.csv", header=True)


and_(less_than_or_equal(mul(x[3],-1),x[0]),or_(x[6],or_(x[8],or_(x[8],and_(x[1],or_(or_(x[6],and_(x[2],or_(x[6],and_(x[9],x[1])))),and_(less_than_or_equal(x[5],sub(x[5],sub(x[0],x[3]))),and_(x[1],less_than_or_equal(mul(x[5],-1.5),x[0])))))))))
Training Fitness:  0.25366917889726304
(14, 95804)
Predicted classes of the training set:  [0 1 0 ... 0 1 0]
Accuracy of the training set:  0.746330821102737
Predicted classes of the test set:  [1 0 0 ... 0 1 0]
    output
0        1
1        0
2        0
3        0
4        1
..     ...
95       0
96       0
97       0
98       0
99       0

[100 rows x 1 columns]


Write a code to create a .csv with the following format:
1. First column is the index (from 0 to 95803);
2. Second column is named `output` and contains the predictions (only 0's or 1's) you  got in the previous cell with y_pred.

Example:

    index,output

    0,0

    1,0

    2,1

    ...

    95803,0


Submit it to the competition and check your score there.